edgraph/server.go

```diff
@@ -1883,6 +1884,356 @@ func (s *Server) CheckVersion(ctx context.Context, c *api.Check) (v *api.Version
+		if !strings.Contains(err.Error(), "error deleting non-existing namespace") {
+			return nil, err
+		} else {
+			glog.Infof("Namespace with id [%d] does not exist, cannot be deleted", nsID)
+		}
+	}
+
+	glog.Infof("Dropped namespace [%v] with id [%d]", in.NsName, nsID)
+	return &apiv25.DropNamespaceResponse{}, nil
+}
+
+func (s *Server) RenameNamespace(ctx context.Context, in *apiv25.RenameNamespaceRequest) (
+	*apiv25.RenameNamespaceResponse, error) {
+
+	if err := AuthGuardianOfTheGalaxy(ctx); err != nil {
+		s := status.Convert(err)
+		return nil, status.Error(s.Code(),
+			"Non guardian of galaxy user cannot rename a namespace. "+s.Message())
+	}
+
+	if err := verifyDropNamespaceName(in.FromNs); err != nil {
+		return nil, err
+	}
+	if err := verifyNamespaceName(in.ToNs); err != nil {
+		return nil, err
+	}
+
+	if strings.HasPrefix(in.FromNs, "dgraph-") {
+		// Only possible when the namespace is created through GraphQL or for pre v25 namespaces.
+		ns, err := strconv.ParseUint(in.FromNs[7:], 10, 64)
+		if err != nil {
+			return nil, errors.Errorf("invalid namespace name %q", in.FromNs)
+		}
+
+		if _, err := getNamespaceName(x.AttachJWTNamespace(ctx), ns); err == nil {
+			// We found the namespace with a real name
+			return nil, fmt.Errorf("namespace [%v] not found: %v", err, in.FromNs)
+		} else if !strings.Contains(err.Error(), "not found") {
+			return nil, err
+		}
+
+		if err := addNamespaceName(x.AttachJWTNamespace(ctx), in.ToNs, ns); err != nil {
+			return nil, err
+		}
+
+		return &apiv25.RenameNamespaceResponse{}, nil
+	}
+
+	resp, err := (&Server{}).QueryNoGrpc(context.WithValue(ctx, IsGraphql, true), &api.Request{
+		Query: `{q as var(func: eq(dgraph.namespace.name, "` + in.FromNs + `"))}`,
+		Mutations: []*api.Mutation{{
+			Cond: `@if(eq(len(q), 1))`,
+			Set: []*api.NQuad{
+				{
+					Subject:     "uid(q)",
+					Predicate:   "dgraph.namespace.name",
+					ObjectValue: &api.Value{Val: &api.Value_StrVal{StrVal: in.ToNs}},
+				},
+			},
+		}},
+		CommitNow: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.Metrics.NumUids["q"] != 1 {
+		return nil, errors.Errorf("namespace [%v] not found", in.FromNs)
+	}
+
+	glog.Infof("Renamed namespace [%v] to [%v]", in.FromNs, in.ToNs)
+	return &apiv25.RenameNamespaceResponse{}, nil
+}
+
+func (s *Server) ListNamespaces(ctx context.Context, in *apiv25.ListNamespacesRequest) (
+	*apiv25.ListNamespacesResponse, error) {
+
+	if err := AuthGuardianOfTheGalaxy(ctx); err != nil {
+		s := status.Convert(err)
+		return nil, status.Error(s.Code(),
+			"Non guardian of galaxy user cannot list namespaces. "+s.Message())
+	}
+
+	const q = `{q(func: type(dgraph.namespace)) {
+			dgraph.namespace.name
+			dgraph.namespace.id }}`
+	resp, err := (&Server{}).doQuery(x.AttachJWTNamespace(ctx),
+		&Request{req: &api.Request{Query: q}, doAuth: NoAuthorize})
+	if err != nil {
+		return nil, err
+	}
+	var data struct {
+		NsList []struct {
```

Comment:
Lets just make a global structure. Why are we declaring it again and again.

Reply:
Done